File: android/app/src/main/java/com/hopchenko/parkpass/ui/detail/Confetti.kt
package com.hopchenko.parkpass.ui.detail

import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.roundToInt
import kotlin.math.sin

private val COLORS = listOf(
    Color(0xFFC67139), Color(0xFF7A8A5E), Color(0xFFF6A06B),
    Color(0xFFAEBF92), Color(0xFF8C491A), Color(0xFFE1EECC),
)

/** CSS `ease-out`. */
private val EaseOut = CubicBezierEasing(0f, 0f, 0.58f, 1f)

private class Dot(i: Int) {
    val size = 6f + (i % 4) * 2f
    val round = i % 2 == 1
    val color = COLORS[i % 6]
    private val angle = i / 16.0 * PI * 2 + (i % 3) * 0.21
    private val dist = 78 + (i % 5) * 14
    val dx = (cos(angle) * dist).roundToInt().toFloat()
    val dy = (sin(angle) * dist - 24).roundToInt().toFloat()
    val duration = 0.65f + (i % 4) * 0.09f
    val delay = (i % 5) * 0.03f
}

/** 16 particles, laid out exactly as ConfettiBurst.tsx — docs/specs/park-detail.md. */
private val DOTS = List(16) { Dot(it) }

/** Total run time in seconds, longest dot included. */
const val CONFETTI_SECONDS = 1.05f

/**
 * Confetti radiating from 45% down the pin area. [elapsed] is seconds since
 * the burst started; before its delay each dot sits at the origin, as CSS
 * `animation-fill-mode: both` does.
 */
@Composable
fun Confetti(elapsed: Float, modifier: Modifier = Modifier) {
    Canvas(modifier) {
        val origin = Offset(size.width / 2f, size.height * 0.45f)
        for (d in DOTS) {
            val local = ((elapsed - d.delay) / d.duration).coerceIn(0f, 1f)
            val e = EaseOut.transform(local)
            val scale = 1f - 0.85f * e
            val alpha = 1f - e
            if (alpha <= 0f) continue
            val side = d.size.dp.toPx() * scale
            val center = origin + Offset(d.dx.dp.toPx() * e, d.dy.dp.toPx() * e)
            val topLeft = center - Offset(side / 2f, side / 2f)
            if (d.round) {
                drawCircle(d.color, radius = side / 2f, center = center, alpha = alpha)
            } else {
                drawRoundRect(
                    d.color,
                    topLeft = topLeft,
                    size = Size(side, side),
                    cornerRadius = CornerRadius(3.dp.toPx() * scale),
                    alpha = alpha,
                )
            }
        }
    }
}

File: ios/ParkPass/Screens/Confetti.swift
import SwiftUI

/// 16 particles, laid out exactly as ConfettiBurst.tsx — docs/specs/park-detail.md.
struct Confetti: View {
    /// Seconds since the burst started.
    let elapsed: TimeInterval

    private struct Dot {
        let size: Double
        let round: Bool
        let color: Color
        let dx: Double
        let dy: Double
        let duration: Double
        let delay: Double

        init(_ i: Int) {
            let colors: [UInt32] = [0xC67139, 0x7A8A5E, 0xF6A06B, 0xAEBF92, 0x8C491A, 0xE1EECC]
            let angle = Double(i) / 16 * .pi * 2 + Double(i % 3) * 0.21
            let dist = Double(78 + (i % 5) * 14)
            size = Double(6 + (i % 4) * 2)
            round = i % 2 == 1
            color = Color(hex: colors[i % 6])
            dx = (cos(angle) * dist).rounded()
            dy = (sin(angle) * dist - 24).rounded()
            duration = 0.65 + Double(i % 4) * 0.09
            delay = Double(i % 5) * 0.03
        }
    }

    private static let dots = (0..<16).map(Dot.init)

    /// CSS `ease-out` = cubic-bezier(0,0,.58,1).
    private static let easeOut = UnitCurve.bezier(startControlPoint: UnitPoint(x: 0, y: 0), endControlPoint: UnitPoint(x: 0.58, y: 1))

    var body: some View {
        Canvas { context, size in
            // Radiates from 45% down the pin area; before its delay each dot
            // waits at the origin, as `animation-fill-mode: both` does.
            let origin = CGPoint(x: size.width / 2, y: size.height * 0.45)
            for d in Self.dots {
                let local = min(1, max(0, (elapsed - d.delay) / d.duration))
                let e = Self.easeOut.value(at: local)
                let alpha = 1 - e
                guard alpha > 0 else { continue }
                let side = d.size * (1 - 0.85 * e)
                let center = CGPoint(x: origin.x + d.dx * e, y: origin.y + d.dy * e)
                let rect = CGRect(x: center.x - side / 2, y: center.y - side / 2, width: side, height: side)
                let shape = d.round
                    ? Path(ellipseIn: rect)
                    : Path(roundedRect: rect, cornerRadius: 3 * (1 - 0.85 * e))
                context.fill(shape, with: .color(d.color.opacity(alpha)))
            }
        }
    }
}
